Carto credentials

tuxskar opened the pull request:
Created a new version of how to deal with the carto credentials using an object

This object provides an API for:
* Getting the credentials dict object
* Getting the cw credentials
* Getting the big query client

It includes a way to save the generated token into the file system to reuse it while it is still valid and all the examples and tests updated

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+        now_utc_ts = datetime.datetime.utcnow().timestamp()
+        return now_utc_ts > self.expiration_ts
+
+    def cw_credentials(self) -> tuple:
```
Rename to `get_carto_dw_credentials`

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+            json.dump(info_to_cache, fw)
+        return True
+
+    def _get_file_token(self):
```
Would be more like `_load_file_token`

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+            self.scope = info["scope"]
+        return True
+
+    def get_bigquery_client(self):
```
Rename to `get_carto_dw_client`

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+        project_id, cw_token = self.cw_credentials()
+        return GClient(project_id, credentials=GCredentials(cw_token))
+
+    def connections_list(self) -> list:
```
I wouldn't include validation functions as part of this PR.

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/examples/connector_bigquery.py:
```diff
@@ -0,0 +1,22 @@
+
+Render cloud data from a query.
+"""
+from pydeck_carto.carto_auth.auth import CartoAuth
```
```
from pydeck_carto import CartoAuth
```

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/examples/connector_bigquery.py:
```diff
@@ -0,0 +1,22 @@
+
+ca = CartoAuth("./carto_credentials.json")
+
+bq_client = ca.get_bigquery_client()
```
cdw_client = ca.get_carto_dw_client()

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/examples/connector_bigquery.py:
```diff
@@ -0,0 +1,22 @@
+for table in table_list:
+    print(table)
+
+datasets = bq_client.list_datasets()
```
List datasets first, list tables second

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/tests/test_bq_client.py:
```diff
@@ -0,0 +1,54 @@
+from pydeck_carto.carto_auth.auth import CartoAuth
+from google.cloud.bigquery import Client as GClient
+
+dummy_response = {
```
Use better mocks:

```
"projectId": "project-id-mock",
"token": "token-mock",
```

So they can be used in the asserts.

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/tests/test_bq_client.py:
```diff
@@ -0,0 +1,54 @@
+    assert hasattr(bq_client, "query")
+    assert isinstance(bq_client, GClient)
+
+    # table_list = bq_client.list_tables("shared")
```
Remove commented code

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+            "accessToken": oauth_token,
+        }
+
+    def get_token(self):
```
Rename to `_get_token` (private api)

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+        return True
+
+    def get_bigquery_client(self):
+        """Returns a client to query directly big query"""
```
Returns a client to query directly the CARTO DW (BigQuery)

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/examples/connector_bigquery.py:
```diff
@@ -0,0 +1,22 @@
+CartoLayer
+==========
+
+Render cloud data from a query.
```
```
Use CARTO DW client (BigQuery)
```

Jesus89 commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+    pass
+
+
+class CartoAuth:
```
Missing class doctring

Jesus89 reviewed (commented):
Overall it looks good.

I have added several naming changes to match the original API, and some WIP code to clean up in this PR.

I'm thinking about providing a more standard way of loading the creds file, similar to [this](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client), and removing `filename` from the constructor:

```py
ca = CartoAuth.from_file("./carto_credentials.json")
```

tuxskar commented on bindings/pydeck-carto/pydeck_carto/carto_auth/auth.py:
```diff
@@ -0,0 +1,217 @@
+        cache_filepath=".carto_token.json",
+        access_token=None,
+        expire_in=None,
+        scope=None,
```
I think that the `cache_filepath` should be kept as it could be configured by the user, maybe to have the cache in a different path, actually one of the tests is using different cache_filepaths to enforce this idea

Jesus89 replied:
OK, let's keep it then

tuxskar commented:
Modified the pull request with the feedback except for removing the `cache_filepath` parameter as I see it quite useful 

Also added a check when using the cached information to make sure the scope is the same requested

tuxskar commented:
Added the missing feedback and checked that the lint is happy back again 😄

Jesus89 commented on bindings/pydeck-carto/examples/carto_layer.ipynb:
```diff
@@ -46,7 +46,7 @@
     "    type_=MapType.QUERY,\n",
     "    connection=CartoConnection.CARTO_DW,\n",
-    "    credentials=credentials,\n",
+    "    credentials=carto_auth.credentials(),\n",
```
get_layer_credentials()

Jesus89 commented on bindings/pydeck-carto/requirements_dev.txt:
```diff
@@ -1,8 +1,9 @@
+-r requirements.txt
```
requirements.txt does not exist

Jesus89 reviewed (approved):
LGTM
